packages/excubiae/contracts/extensions/SemaphoreExcubia.sol

```diff
@@ -0,0 +1,78 @@
+/// To avoid identities from passing twice, nullifiers are stored upon successful verification of the proofs.
+contract SemaphoreExcubia is Excubia {
+    /// @notice The Semaphore contract interface.
+    ISemaphore public immutable SEMAPHORE;
+    /// @notice The specific group identifier that proofs must match to pass the gate.
+    /// @dev Used as a `scope` to ensure consistency during proof membership verification.
+    uint256 public immutable GROUP_ID;
+
+    /// @notice Mapping to track which nullifiers have been used to avoid passing the
+    /// gate twice using the same Semaphore identity.
+    /// @dev The nullifier is derived from the hash of the secret and group identifier,
+    /// ensuring that the same identity cannot be registered twice for the same group.
+    mapping(uint256 => bool) public passedNullifiers;
+
+    /// @notice Error thrown when the group identifier does not match the expected one.
+    error InvalidGroup();
+
+    /// @notice Error thrown when the proof is invalid.
+    error InvalidProof();
+
+    /// @notice Error thrown when the proof scope does not match the expected group identifier.
+    error UnexpectedScope();
+
+    /// @notice Constructor to initialize with target Semaphore contract and specific group identifier.
+    /// @param _semaphore The address of the Semaphore contract.
+    /// @param _groupId The group identifier that proofs must match.
+    constructor(address _semaphore, uint256 _groupId) {
+        if (_semaphore == address(0)) revert ZeroAddress();
+
+        SEMAPHORE = ISemaphore(_semaphore);
+
+        if (ISemaphore(_semaphore).groupCounter() < _groupId) revert InvalidGroup();
+
+        GROUP_ID = _groupId;
+    }
+
+    /// @notice Internal function to handle the passing logic with check.
+    /// @dev Calls the parent `_pass` function and registers the nullifier to avoid passing the gate twice.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    /// @param data Additional data required for the check (ie., encoded Semaphore proof).
+    function _pass(address passerby, bytes calldata data) internal override {
+        ISemaphore.SemaphoreProof memory proof = abi.decode(data, (ISemaphore.SemaphoreProof));
+
+        // Avoiding passing the gate twice using the same nullifier.
+        if (passedNullifiers[proof.nullifier]) revert AlreadyPassed();
+
+        super._pass(passerby, data);
+
+        passedNullifiers[proof.nullifier] = true;
+    }
+
+    /// @notice Internal function to handle the gate protection (proof check) logic.
+    /// @dev Checks if the proof matches the group ID, scope, and is valid.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    /// @param data Additional data required for the check (i.e., encoded Semaphore proof).
+    /// @return True if the proof is valid and the passerby passes the check, false otherwise.
```

Comment:
Looks like this function never returns false. I may not have enough context on this function but if it is intended to return false there shouldn't be reverts here. It should be something like:

```solidity
return GROUP_ID == proof.scope && SEMAPHORE.verifyProof(GROUP_ID, proof)
```
(if `super._check(passerby, data)` returns a bool it should also be included maybe).

Otherwise, that comment needs to be updated.

Please, correct me if I'm wrong.

Reply:
It seems reasonable to eliminate the return value here. Unfortunately, the design does not match the `modifier` syntax since I'd like to have `check` as a method externally callable for Excubia aggregation purposes and separation of concerns. Making it a modifier will result in a non-so-clear design for check-pass logic.

But yes, it's possible to eliminate the return value entirely and use `_check` as a method that if does not revert, check is passed. In fact, we weren't using the `AccessDenied` error since we were always returning true. I will create a pull request for this purpose. I appreciate the feedback <3